mlflow/types/schema.py

```diff
@@ -122,34 +122,267 @@ def from_numpy_type(cls, np_type):
+    def name(self) -> str:
+        """The property name."""
+        return self._name
+
+    @property
+    def dtype(self) -> Union[DataType, ArrayType, ObjectType]:
+        """The property data type."""
+        return self._dtype
+
+    @property
+    def required(self) -> bool:
+        """Whether this property is required"""
+        return self._required
+
+    @required.setter
+    def required(self, value: bool) -> None:
+        self._required = value
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Property):
+            return self.dtype == other.dtype and self.required == other.required
+        return False
+
+    def to_dict(self):
+        d = {"name": self.name}
+        if isinstance(self.dtype, DataType):
+            d["type"] = self.dtype.name
+        else:
+            d.update(self.dtype.to_dict())
+        d["required"] = self.required
+        return d
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `name`, `type` and
+        optional `required` keys.
+        """
+        if not {"name", "type"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in Property JSON. Expected to find keys `name` and `type`"
+            )
+        name = kwargs.pop("name")
+        required = kwargs.pop("required", True)
+        dtype = kwargs["type"]
+        if dtype not in ["array", "object"]:
+            return cls(name=name, dtype=dtype, required=required)
+        if dtype == "array":
+            return cls(name=name, dtype=Array.from_json_dict(**kwargs), required=required)
+        if dtype == "object":
+            return cls(name=name, dtype=Object.from_json_dict(**kwargs), required=required)
+
+
+@experimental
+class Object:
+    """
+    Specification used to represent a json-convertible object.
+    """
+
+    def __init__(self, properties: List[Property]) -> None:
+        self._check_properties(properties)
+        self._properties = properties
+
+    def _check_properties(self, properties):
+        if not isinstance(properties, list):
+            raise MlflowException.invalid_parameter_value(
+                f"Expected properties to be a dict, got type {type(properties).__name__}"
+            )
+        if any(not isinstance(v, Property) for v in properties):
+            raise MlflowException.invalid_parameter_value(
+                "Expected values to be instance of Property"
+            )
+        # check duplicated property names
+        names = set()
+        duplicates = set()
+        for prop in properties:
+            if prop.name in names:
+                duplicates.add(prop.name)
+            else:
+                names.add(prop.name)
+        if len(duplicates) > 0:
+            raise MlflowException.invalid_parameter_value(
+                f"Found duplicated property names: {duplicates}"
+            )
```

Comment:
Thoughts on a set comprehension?
```suggestion
        names = {prop.name for prop in properties}
        duplicates = names.intersection({prop.name for prop in properties if 
          properties.count(prop.name) > 1})

        if duplicates:
            raise MlflowException.invalid_parameter_value(
                f"Found duplicated property names: {duplicates}"
            )
```